### PR TITLE
Citizen reservation fixes

### PR DESCRIPTION
**Pull request by akheron**
#### Summary

* Don't allow citizen to add reservations or absences if staff has marked absences.
* Make some frontend optimizations while at it.
* Add child pictures to the calendar and the day modal

Desktop calendar:

Screenshot 2022-03-28 at 7 51 22

Mobile calendar:

Screenshot 2022-03-28 at 7 51 52

Day modal:

Screenshot 2022-03-28 at 7 52 21

### CODE REVIEW

File: frontend/src/citizen-frontend/calendar/RoundChildImages.tsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+const ChildImage = styled.img<{ size: number; border: number }>`
+  ${roundMixin};
+`
+
+const accentColors = [
+  theme.colors.accents.a6turquoise,
+  theme.colors.accents.a9pink,
+  theme.colors.accents.a7mint,
+  theme.colors.accents.a5orangeLight
+]
```
Comment (kimifoud) [inline frontend/src/citizen-frontend/calendar/RoundChildImages.tsx:106]: Pretty sure these need to be customizable, but this is fine for now.

File: frontend/src/citizen-frontend/calendar/RoundChildImages.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    .map((child, index) => [child, index] as const)
+    .filter(([child, _]) =>
+      day.children.some(
+        (dc) =>
+          dc.childId === child.id &&
+          dc.absence === null &&
+          (dc.reservations.length > 0 || dc.attendances.length > 0)
+      )
+    )
+    .map(([child, index]) => ({
```
Comment (kimifoud) [inline frontend/src/citizen-frontend/calendar/RoundChildImages.tsx:135]: Couldn't this be simplified by filtering before adding the index?

File: frontend/src/citizen-frontend/calendar/CalendarGridView.tsx [hunk 3]
```diff
@@ -327,29 +491,32 @@ const DayCell = styled.button<{
...
+      : ''};
 
   ${(p) =>
     p.selected
       ? css`
           box-shadow: 0 2px 3px 2px #00000030;
           z-index: 1;
         `
-      : ''}
+      : ''};
```
Comment (kimifoud) [inline frontend/src/citizen-frontend/calendar/CalendarGridView.tsx:519]: Redundant semicolons? 🤔

File: service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt [hunk 4]
```diff
@@ -67,9 +67,20 @@ fun Database.Transaction.deleteAbsencesCreatedFromQuestionnaire(questionnaireId:
...
+ */
+fun Database.Transaction.clearOldCitizenEditableAbsences(childDatePairs: List<Pair<ChildId, LocalDate>>) {
     val batch = prepareBatch(
-        "DELETE FROM absence WHERE child_id = :childId AND date = :date AND absence_type <> 'FREE_ABSENCE'::absence_type"
+        """
+DELETE FROM absence 
+WHERE child_id = :childId
+AND date = :date
+AND absence_type <> 'FREE_ABSENCE'::absence_type
+AND (SELECT evaka_user.type = 'CITIZEN' FROM evaka_user WHERE evaka_user.id = absence.modified_by)
```
Comment (kimifoud) [inline service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt:82]: The same query is used when employee creates reservations for children in the week view. So, it effectively prevents creating a reservation when there is an absence on the same day. The save operation just silently ignores the saved attendance unless the absence is deleted first. This seems fine to me, but the UX is slightly confusing at first.
kuva

**Review by kimifoud — APPROVED** [review]
🥇

File: frontend/src/citizen-frontend/calendar/RoundChildImages.tsx (hunk 2 above)
Comment (akheron) [reply frontend/src/citizen-frontend/calendar/RoundChildImages.tsx:135]: No. If the child doesn't have a picture, their placeholder color is selected based on the index. Whether or not the child happens to have a reservation for this particular day should not affect the color.

Added a comment that clarifies this 👍

File: frontend/src/citizen-frontend/calendar/CalendarGridView.tsx (hunk 3 above)
Comment (akheron) [reply frontend/src/citizen-frontend/calendar/CalendarGridView.tsx:519]: Yeah kinda, but IntelliJ IDEA's autoformat breaks without them and styled-components seems to handle them correctly.

File: service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt (hunk 4 above)
Comment (akheron) [reply service/src/main/kotlin/fi/espoo/evaka/reservations/ReservationQueries.kt:82]: Let's check this UX separately

**Pull request merged**